org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -427,18 +433,74 @@ protected void compile(SourceFile[] units, SourceFile[] additionalUnits, boolean
 		for (int i = 0; i < toAdd; i++)
 			additionalUnits[length + i] = iterator.next();
 	}
-	String[] initialTypeNames = new String[units.length];
-	for (int i = 0, l = units.length; i < l; i++) {
-		char[] moduleName = units[i].getModuleName();
-		initialTypeNames[i] = (moduleName == null)
-				? units[i].initialTypeName
-				: new StringBuilder(60).append(moduleName).append(':').append(units[i].initialTypeName).toString();
-	}
-	this.nameEnvironment.setNames(initialTypeNames, additionalUnits);
 	this.notifier.checkCancel();
+	boolean warnAboutMissingReleaseFlag = false;
 	try {
 		this.inCompiler = true;
-		this.compiler.compile(units);
+		Map<Integer, List<SourceFile>> collect = Arrays.stream(units)
+				.collect(Collectors.groupingBy(sf -> sf.sourceLocation.release, TreeMap::new, Collectors.toList()));
+		for (Entry<Integer, List<SourceFile>> entry : collect.entrySet()) {
+			long oldTarget = this.compiler.options.targetJDK;
+			long oldCompliance = this.compiler.options.complianceLevel;
+			long oldSource = this.compiler.options.sourceLevel;
+			boolean oldRelease = this.compiler.options.release;
+			INameEnvironment oldEnv = this.compiler.lookupEnvironment.nameEnvironment;
+			try {
+				int release = entry.getKey();
+				if (release >= JavaProject.FIRST_MULTI_RELEASE) {
+					long currentTarget = CompilerOptions.releaseToJDKLevel(release);
+					this.compiler.options.targetJDK = currentTarget;
+					this.compiler.options.complianceLevel = currentTarget;
+					this.compiler.options.sourceLevel = currentTarget;
+					this.compiler.options.release = true;
+					try {
+						this.compiler.lookupEnvironment.nameEnvironment = this.javaBuilder.getNameEnvironment(release);
+					} catch (CoreException e) {
+						List<IContainer> list = entry.getValue().stream().map(sf -> sf.sourceLocation.sourceFolder)
+								.distinct().toList();
+						for (IContainer container : list) {
+							createProblemFor(container, null,
+									NLS.bind(Messages.AbstractImageBuilder_env_failed,
+											new Object[] { release, e.getStatus()}),
```

Comment:
thanks for trying to handle this exception, but does it really signal something that warrants a problem marker? Certainly "LookupEnvironment" is not a concept that should be mentioned in user-facing messages.
Unfortunately, the location where the exception is raised does not give any hint on why it can occur.

Reply:
`org.eclipse.jdt.internal.core.builder.JavaBuilder.initializeBuilder(int, boolean)` init the name environment that then is shared with the `AbstractImageBuilder` ... I could init them all there as well but though it might be better to be a bit more lazy here.

Reply:
OK, throwing `ImageBuilderInternalException` makes exception handling look more like existing code, but I don't think that the exception thrown here will be caught somewhere in every possible call path. E.g., we may be indirectly called from `IncrementalImageBuilder.incrementalBuildLoop()` where I don't see any catching of this particular exception in place. Since it's a RuntimeException it's your responsibility to ensure proper catching :)

To me it would look safer if the exception is indeed handled here, but in a way that re-uses or copies the code combining logging and a specific buildpath error marker.

Reply:
I found at least three different prior-arts in this class ... seems I decided for the wrong one, I now used your recommended approach.